Python/Product/Analysis/LanguageServer/ParseQueue.cs

```diff
@@ -0,0 +1,209 @@
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.PythonTools.Analysis.Infrastructure;
+using Microsoft.PythonTools.Intellisense;
+using Microsoft.PythonTools.Parsing;
+using Microsoft.PythonTools.Parsing.Ast;
+
+namespace Microsoft.PythonTools.Analysis.LanguageServer {
+    class ParseQueue {
+        public const string PythonParserSource = "Python";
+        private const string TaskCommentSource = "Task comment";
+
+        private readonly ConcurrentDictionary<IDocument, TaskCompletionSource<IAnalysisCookie>> _parsing;
+
+        private readonly VolatileCounter _parsingInProgress;
+
+        public ParseQueue() {
+            _parsingInProgress = new VolatileCounter();
+            _parsing = new ConcurrentDictionary<IDocument, TaskCompletionSource<IAnalysisCookie>>();
+        }
+
+        public int Count => _parsingInProgress.Count;
+
+        public Task WaitForAllAsync() => _parsingInProgress.WaitForZeroAsync();
+
+        private sealed class ParseState {
+            public IDocument Document;
+            public PythonLanguageVersion LanguageVersion;
+            public TaskCompletionSource<IAnalysisCookie> Task;
+        }
+
+        private static void AbortParsingTree(IPythonProjectEntry entry) {
+            if (entry == null) {
+                return;
+            }
+
+            entry.GetTreeAndCookie(out var tree, out var cookie);
+            entry.UpdateTree(tree, cookie);
+        }
+
+        internal bool TryGetExistingParseAsync(IDocument doc, out Task<IAnalysisCookie> task) {
+            if (_parsing.TryGetValue(doc, out var tcs)) {
+                task = tcs.Task;
+                return true;
+            }
+            task = null;
+            return false;
+        }
+
+        public async Task<IAnalysisCookie> Enqueue(IDocument doc, PythonLanguageVersion languageVersion) {
+            TaskCompletionSource<IAnalysisCookie> created = null;
+            var tcs = _parsing.GetOrAdd(doc, e => { return created = new TaskCompletionSource<IAnalysisCookie>(); });
+            if (created == null) {
+                // Do not start parsing until the previous one has completed
+                await tcs.Task;
+                tcs = new TaskCompletionSource<IAnalysisCookie>();
```

Comment:
If this method is called three times in a row for the same doc, second and third calls will await for the first one and then will run concurrently. You probably need an async ARE here.

Reply:
Or you can simplify this by using `AddOrUpdate`
```cs
public Task<IAnalysisCookie> Enqueue(IDocument doc, PythonLanguageVersion languageVersion) {
    return _parsing.AddOrUpdate(doc, 
        d => Task.Run(() => ParseWorker(d, languageVersion)),
        (d, task) => task.ContinueWith(t => ParseWorker(d, languageVersion)));
}
```
and get rid of the `TaskCompletionSource` altogether.

Reply:
Oh that's nice. I didn't know you got the previous value in AddOrUpdate.